[TECH] Utilisation de npm workspaces.

HEYGUL opened the pull request:
## :unicorn: Problème

Les targets `npm` à rallonge, les `npm run ci:all`, etc .... On n'en (p|v)eut plus !

## :robot: Proposition

Utiliser `npm workspaces` pour tirer parti du monorepo.
Dans le cas de l'utilisation de workspaces, les surchages de dépendances doivent être déclarées dans le `package.json` racine, ce qui entraine la copie des paramètres `overrides` de chaque package dans le `package.json` racine.

## :rainbow: Remarques

Toutes les applications sauf `pix-audit-logger` et `1d` ont été migrées dans un workspace.
Ces deux applications posent des soucis à la résolution des dépendances de `npm`. On les laisse de côté pour le moment. Pour `pix-audit-logger` nous utilisons node 18 au lieu de node 16.

Concernant le déploiement sur scalingo, celui-ci a été adapté pour les Review apps et les applications type production (intégration / recette / production).
Pour déployer en mode application unique (soit un conteneur = une application front), il est nécessaire de définir la variable d'environnement `USE_NPM_INSTALL=false` pour forcer l'utilisation de la commande `npm ci`.


## :100: Pour tester

Une CI verte, what else ?

En local, supprimez tous vos node_modules avec par exemple `npx npkill .` puis faire un `npm install` à la racine.

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6791.review.pix.fr)
  * [App (.org)](https://app-pr6791.review.pix.org)
  * [Orga (.fr)](https://orga-pr6791.review.pix.fr)
  * [Orga (.org)](https://orga-pr6791.review.pix.org)
  * [Certif (.fr)](https://certif-pr6791.review.pix.fr)
  * [Certif (.org)](https://certif-pr6791.review.pix.org)
  * [Admin](https://admin-pr6791.review.pix.fr)
  * [API](https://api-pr6791.review.pix.fr/api/)
  * [Pix1D](https://1d-pr6791.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6791/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6791/environment)

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6791.review.pix.fr)
  * [App (.org)](https://app-pr6791.review.pix.org)
  * [Orga (.fr)](https://orga-pr6791.review.pix.fr)
  * [Orga (.org)](https://orga-pr6791.review.pix.org)
  * [Certif (.fr)](https://certif-pr6791.review.pix.fr)
  * [Certif (.org)](https://certif-pr6791.review.pix.org)
  * [Admin](https://admin-pr6791.review.pix.fr)
  * [API](https://api-pr6791.review.pix.fr/api/)
  * [Pix1D](https://1d-pr6791.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6791/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6791/environment)

HEYGUL committed:
tech(orga): move orga to npm workspaces

pix-bot-github commented:
Une fois les applications déployées, elles seront accessibles via les liens suivants :
  * [App (.fr)](https://app-pr6791.review.pix.fr)
  * [App (.org)](https://app-pr6791.review.pix.org)
  * [Orga (.fr)](https://orga-pr6791.review.pix.fr)
  * [Orga (.org)](https://orga-pr6791.review.pix.org)
  * [Certif (.fr)](https://certif-pr6791.review.pix.fr)
  * [Certif (.org)](https://certif-pr6791.review.pix.org)
  * [Admin](https://admin-pr6791.review.pix.fr)
  * [API](https://api-pr6791.review.pix.fr/api/)
  * [Pix1D](https://1d-pr6791.review.pix.fr)

Les variables d'environnement seront accessibles via les liens suivants :
  * [scalingo front](https://dashboard.scalingo.com/apps/osc-fr1/pix-front-review-pr6791/environment)
  * [scalingo api](https://dashboard.scalingo.com/apps/osc-fr1/pix-api-review-pr6791/environment)

HEYGUL committed:
tech: revert "Upgrade to ember-cli-mirage 3.0.0-alpha.3"

This reverts commit fc4442a8021fb94227372df6e4c35fd2caf5eccb.

HEYGUL committed:
tech(admin): move admin to npm workspaces

HEYGUL committed:
tech(mon-pix): move mon-pix to npm workspaces

HEYGUL committed:
tech(certif): move certif to npm workspaces

HEYGUL committed:
tech: use workspaces for building

HEYGUL committed:
tech: remove now uneeded CI cache for workspaced apps

HEYGUL committed:
tech(api): move api to npm workspaces

HEYGUL committed:
chore: remove now useless run scripts

Switching to workspaces gives the advantage to use the `--workspace=`
option of npm to directly run a script in the context of decicated
package.

HEYGUL committed:
chore: fix prettier on certif

HEYGUL commented on package.json:
```diff
@@ -22,28 +22,18 @@
   },
   "main": "index.js",
   "scripts": {
+    "prebuild": "mkdir -p ../dist",
```
```suggestion
    "prebuild": "mkdir -p ./dist",
```

HEYGUL commented on 1d/package.json:
```diff
@@ -18,6 +18,7 @@
   },
   "files": [],
   "scripts": {
+    "scalingo-prebuild": "mv ../package-lock.json package-lock.json; rm ../package.json",
```
A supprimer

annemarie35 commented on admin/package.json:
```diff
@@ -19,7 +19,7 @@
   },
   "scripts": {
-    "scalingo-prebuild": "mv ../package-lock.json package-lock.json; rm ../package.json",
+    "scalingo-prebuild": "rm ../node_modules/mon-pix; cp -rL ../node_modules ./node_modules; rm ../package.json",
```
ça devrait pas être `"rm ../node_modules/admin ` ?

octo-topi reviewed (dismissed):
Je ne sais pas où mettre ça en review, donc je le mets là !

Deux points me frappent:
- on ne sait pas pourquoi on le fait (quels avantages ? quels inconvénients ?) : je m'attends à un ADR car c'est structurant (ex: on a déjà mis le choix des versions de Node dans un ADR)
- on ne voit pas d'impact dans la doc d'installation, est-ce normal ? Rien ne change ? Vraiment ? On fait toujours `npm ci` dans chaque app ?

annemarie35 commented:
@octo-topi on est en train de mettre tout ça dans les slides de présentation, l'adr est une bonne idée ^^

francois2metz commented:
Avec le fonctionnement de npm et de la résolution de dépendance, le fait de devoir synchroniser les versions des packages entre les applis front me parait bloquant.

Thomasevano committed:
docs(adr): add adr related to npm workspaces
Co-authored-by: Anne-Marie Esteves <anne-marie.esteves@octo.com>
Co-authored-by: Ismael Gorissen <ismael.gorissen@gmail.com>

Co-authored-by: Guillaume Lagorce <guillaume.lagorce@pix.fr>
Co-authored-by: Emeric Martineau <emeric.martineau@pix.fr>

HEYGUL commented:
> Avec le fonctionnement de npm et de la résolution de dépendance, le fait de devoir synchroniser les versions des packages entre les applis front me parait bloquant.

Pas si sûr : est-ce que renovate saurait mettre à jour les dépendances de manière synchrone sur les applis front.
On pourrait ainsi bénéficier de ce premier travail avant de passer à une autre solution (genre `pnpm`).

yannbertrand commented:
Quelques petites découverte pour moi.

Je ne connais pas bien les workspaces, j'ai regardé un peu [la doc](https://docs.npmjs.com/cli/v7/using-npm/workspaces) puis j'ai fait quelques tests depuis la racine :
- `npm ci`
- `npm run test:api:unit --workspace api` - pratique, pas besoin de me déplacer dans le dossier `api` !
- `npm run start --workspace api mon-pix` ne lance que l'API, c'est normal d'après la doc
- `npm run start --workspace api --workspace mon-pix` ne lance que l'API. Quand je `ctrl+c`, l'API se coupe et mon-pix se lance donc les 2 ne tournent pas en même temps 😅 peut être que j'ai mal compris le fonctionnement.

Du coup pour lancer mon env de dev pour Pix App, j'ai toujours besoin de 2 terminaux mais je peux les lancer depuis la racine `npm run start -w api` et `npm run start -w mon-pix` dans un autre terminal.

`npm run lint --workspaces` semble paralléliser correctement.

`npm run start -ws` lance en série 🤔, il faut interrompre toutes les apps les unes après les autres pour le couper 😅

Et les commandes habituelles restent fonctionnelles :)

igorissen commented:
> * `npm run start --workspace api mon-pix` ne lance que l'API, c'est normal d'après la doc
> * `npm run start --workspace api --workspace mon-pix` ne lance que l'API. Quand je `ctrl+c`, l'API se coupe et mon-pix se lance donc les 2 ne tournent pas en même temps 😅 peut être que j'ai mal compris le fonctionnement.

Oui on avait fait ces tests aussi

> `npm run lint --workspaces` semble paralléliser correctement.

Derrière ce run script NPM il y a cette commande `run-p --print-label lint:*`

yannbertrand commented:
> > `npm run lint --workspaces` semble paralléliser correctement.
> 
> Derrière ce run script NPM il y a cette commande `run-p --print-label lint:*`

Ah j'ai été un peu vite, en fait non ça tourne en série ! Le fait de préciser `-ws` fait bien en sorte de ne pas exécuter le script à la racine.

yannbertrand commented:
> > Avec le fonctionnement de npm et de la résolution de dépendance, le fait de devoir synchroniser les versions des packages entre les applis front me parait bloquant.
> 
> Pas si sûr : est-ce que renovate saurait mettre à jour les dépendances de manière synchrone sur les applis front. On pourrait ainsi bénéficier de ce premier travail avant de passer à une autre solution (genre `pnpm`).

On peut configurer Renovate pour grouper tout ça, mais Renovate ne fait que proposer les nouvelles versions. Parfois il faut faire du travail manuel derrière et des fois c'est déjà assez difficile unitairement sur une appli. Ça a été plutôt un avantage jusqu'à présent de dissocier les versions des applis pour y aller incrémentalement. Donc je partage la crainte : le risque est de freiner les montées de versions.

Ceci dit je comprend la limitation, si vous pensez que la solution `pnpm` est viable et trouvez un moyen d'y aller (quite à trouver plus de temps pour avec une impact team ?), je suis plutôt OK pour apporter ce changement même si il n'est que temporaire.

Par contre je suis d'accord qu'il manque de précision sur la problématique que cette proposition essaye de résoudre pour le moment ! Ça serait intéressant d'avoir plus d'éléments dans le descriptif de la PR :)

Thomasevano commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,47 @@
+## Contexte
+Décrire l'état actuel
+
+Mais nous afrrivons aux limites :
```
```suggestion
Mais nous arrivons aux limites :
```

HEYGUL committed:
tech: ensure all dep of front workpaces are in a single renovate upgrade PR

igorissen committed:
refactor: update package.json scripts

Co-authored-by: Anne-Marie Esteves <anne-marie.esteves@octo.com>
Co-authored-by: Thomas Evano <thomas.evano@pix.fr>
Co-authored-by: Emeric Martineau <emeric.martineau@pix.fr>

annemarie35 commented on package.json:
```diff
@@ -23,20 +23,26 @@
-    "ci:front": "run-s --print-label ci:all ci:1d",
-    "ci:all": "npm ci",
+    "build:admin": "npm run build --workspace=admin",
+    "build:api": "npm run build --workspace=api",
```
build:api à supprimer

HEYGUL commented on INSTALLATION.md:
```diff
@@ -97,7 +97,7 @@ Pour démarrer l'ensemble des applications, exécuter `npm start `
-Si cela pose problème, démarrer sélectivement les applications :
+Si cela pose un problème, démarrer sélectivement les applications :
 
 - Admin : `npm run start:admin`
```
npm run start --workspace=admin

HEYGUL commented on INSTALLATION.md:
```diff
@@ -97,7 +97,7 @@ Pour démarrer l'ensemble des applications, exécuter `npm start `
+Si cela pose un problème, démarrer sélectivement les applications :
 
 - Admin : `npm run start:admin`
 - Api : `npm run start:api`
```
npm run start --workspace=api

HEYGUL commented on package.json:
```diff
@@ -23,20 +23,26 @@
-    "ci:all": "npm ci",
+    "build:admin": "npm run build --workspace=admin",
+    "build:api": "npm run build --workspace=api",
+    "build:certif": "npm run build --workspace=certif",
+    "build:mon-pix": "npm run build --workspace=mon-pix",
+    "build:orga": "npm run build --workspace=orga",
+    "ci:all": "run-p --print-label ci:1d ci:workspaces",
+    "ci:workspaces": "npm ci",
```
pourquoi les remettre ? perso je préfère que chaque s'approprie l'usage de npm

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+
+## Contexte
+
+Notre utilisation actuelle de NPM n'est pas optimale :
+- Des dépendances dupliquées entre les différentes app
+- Taille des node_modules importante
+- Temps d’installation des dépendances assez long
```
Cette phrase est problématique de mon point de vue  car:
- elle est arbitraire/subjective (`optimale`, `important`, `long`); le but ici est d'examiner objectivement la situation. S'il était évident que les `workspace` sont mieux, on l'aurait fait depuis le début.
- elle mélange la cause et les conséquences
- elle omet un élément de contexte important: tous les environnements ne sont pas éligibles.

Je propose cette reformulation.
```suggestion
Chaque application déclare ses propres dépendances avec `npm` , sans les mettre en commun avec les autres applications. En conséquence, ces dépendances sont installées isolément. Chaque application est autoportante.  

Si les dépendances sont mises en commun (avec les ` workspaces`), elles ne seront téléchargées et installées qu'une seule fois. Cela peut réduire le temps d'installation et la taille prise sur le disque. Cette solution n'est valide qu'en local, pas en production où les applications sont déployées séparément.
```

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+
+**Avantage(s) :**
+
+- Un dossier node_modules à la racine regroupant les dépendances communes des workspaces
```
Ce n'est pas un avantage, c'est un détail d'implémentation.
```suggestion
```

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+**Avantage(s) :**
+
+- Un dossier node_modules à la racine regroupant les dépendances communes des workspaces
+- La taille des dossiers node_modules est divisé par deux
+- Le temps d'installation est divisé par deux
```
Je n'ai pas la référence en tête, mais il y a un piège ici.
Si la taille est de 2Mo sur un PC récent (pas un système emabarqué), passer à 1Mo n'est pas un avantage comparé à la complexité des workspace. Idem si on passe de 2 secondes à 1 seconde.
```suggestion
Sur un PC 8CPU 2Ghz avec débit descendant de 3Mb/s
- la taille est divisée par 2, passant de 2Gb à 1 Gb
- le temps d'installation est divisé par deux, passant de 10 minutes à 5 minutes
```

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+
+**Inconvénient(s) :**
+
+Deux applications n'ont pu être intégrées aux workspaces :
+
+- Pix 1D, le resolver utilise une mauvaise version des peerDependencies d'ember-data
+- Pix Audit Logger n'utilise pas la même version de Node
+
+Les montées de versions doivent être faites sur toutes les apps en même temps or, elles n'ont pas toutes le même cycle de vie.
+Mais ceci est mitigé par l'utilisation de Renovate (lien vers [commit](https://github.com/1024pix/pix/pull/6791/commits/7956548245f6bbfaff21efc4f6ec94672b51aad0)).
```
Je propose de ne garder que cet inconvénient, et de déplacer les remarques sur pix1d dans la partie conséquences en fin de fichier.

Je propose aussi d'expliciter le travail initial de réalignement des dépendances, pas seulement les montées de version suivantes.
```suggestion
Pour que le gain soit effectif, il faut que les applications aient un nombre suffisant de dépendances en commun. On peut s'attendre à ce que les 4 front-end EmberJs aient des librairies en commun, mais beaucoup moins avec le back-end HapiJs.

Même si c'est le cas, pour que la dépendance puisse être utilisée par plusieurs applications, toutes ces applications devront utiliser la même version de cette librairie. Il y aura donc un travail initial d'alignement de toutes les dépendances partagées entre les applications, sinon il n'y aura pas de bénéfice.

Les montées de versions affecteront toutes les applications. Si uen montée de version embarque un changement bloquant, le code de toutes les applications devra être être modifié avant de merger la PR, ce qui augmentera le temps de développement, de revue, et donc le délai de mise à disposition.

Une partie du travail de montées de version est pris en charge par Renovate (lien vers [commit](https://github.com/1024pix/pix/pull/6791/commits/7956548245f6bbfaff21efc4f6ec94672b51aad0)).

En résumé, l'inconvénient est le couplage entre applications.
```

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+- Taille des node_modules importante
+- Temps d’installation des dépendances assez long
+
+### Solution :
```
Il manque la solution 2 (existante), à savoir utiliser la version native (sans workspace).
C'est important pour que le développeur ne doive pas inverser les inconvénients de la solution 1 dans sa tête pour avoir la solution 2, voir l'ADR 1

```suggestion
### Solution 1 : utiliser les workspace npm
```

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+
+## Décision
+
+Nous avons choisi d'utiliser les npm workspaces.
```
L'intérêt d'un ADR est de dire pourquoi on choisi. Souvent, le contexte rend une décision évidente, mais ce contexte change et on se demande après coup pourquoi la décision.

Je propose de suivre le template (ADR1) et d'expliciter ce qui vous a fait décider en faveur de la solution 1
```
Nous avons choisi la solution n°1, à savoir les ADR, car le temps passé à créer un ADR est minimal par rapport au risque encouru si la connaissance est perdue.
```
Ce qui donnerait ici
```suggestion
Nous avons choisi d'utiliser les npm workspaces parce que l'avantage XXXX est plus important que les inconvénients
```

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+
+## Conséquences
+
+- Impact minime sur le workflow.
```
Cette phrase est mystérieuse.. Quel workflow ? Celui de CI ? Les actions du développeur ?
```suggestion
- Impact minime sur le workflow.
```

octo-topi commented on 0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,46 @@
+- Un dossier node_modules à la racine regroupant les dépendances communes des workspaces
+- La taille des dossiers node_modules est divisé par deux
+- Le temps d'installation est divisé par deux
+- Moins de run script npm
```
J'aimerai une réponse détaillée, le fichier INSTALLATION.md me donne des pistes
```
cd admin && npm start
npm start --workspace=admin
```

octo-topi commented on docs/adr/0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,78 @@
+
+- Chaque version des dépendances est installée une seule fois
+
+Sur un PC 4CPU 2,3Ghz avec débit descendant de 676Mb/s
+- La taille des dossiers node_modules est divisé par deux, passant de 3,17 à 1,51 GB
+- Le temps d'installation est divisé par deux, passant de 454s à 209s
+- Moins de run script npm. Avec l'option `--workspace` nous pouvons spécifier le workspace sur lequel exécuter la commande
```
:fire:

octo-topi commented on docs/adr/0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,78 @@
+
+En résumé, l'inconvénient est le couplage entre applications.
+
+### Solution 2 : rester dans l'état actuel
```
Une ADR peut changer la codebase, bien sûr, mais l'essentiel est qu'on sache quelle solution choisir, je propose donc
```suggestion
### Solution 2 : utiliser npm sans les workspace
```

octo-topi commented on docs/adr/0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,78 @@
+
+**Inconvénient(s) :**
+
+Les mêmes versions des dépendances sont dupliquées.
```
```suggestion
Les dépendances utilisées par plusieurs applications, avec la même version, sont dupliquées.
```

yannbertrand replied:
```suggestion
Les dépendances utilisées par plusieurs applications, y compris si elles dépendant de la même version, sont dupliquées.
```

octo-topi commented on docs/adr/0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,78 @@
+
+## Décision
+
+Nous avons choisi d'utiliser les npm workspaces parce que l'avantage d'avoir _chaque version des dépendances installée une seule fois_ est plus important que les inconvénients.
```
Mouais, nan ! Je propose plus direct
```suggestion
Nous avons choisi d'utiliser les npm workspaces parce que, même si cela amène de la complexité et du couplage entre les applications, l'expérience de développement en local (délai installation, taille disque) est plus importante que la production.
```

octo-topi commented on INSTALLATION.md:
```diff
@@ -93,17 +93,11 @@ Les extensions recommandées peuvent se retrouver dans l'onglet extension en ren
-- App : `npm run start:mon-pix`
-- Certif : `npm run start:certif`
-- Orga : `npm run start:orga`
+- Admin : `npm start --workspace=admin`
+- Api : `npm start --workspace=api`
+- App : `npm start --workspace=mon-pix`
+- Certif : `npm start --workspace=certif`
+- Orga : `npm start --workspace=orga`
```
```suggestion

Chaque application front-end a besoin de l'API pour fonctionner.

Pour démarrer:
- Admin : `npm start --workspace=admin`
- Api : `npm start --workspace=api`
- App : `npm start --workspace=mon-pix`
- Certif : `npm start --workspace=certif`
- Orga : `npm start --workspace=orga`
```

yannbertrand commented on docs/adr/0052-utilisation-de-workspaces.md:
```diff
@@ -0,0 +1,78 @@
+
+Chaque application déclare ses propres dépendances avec `npm`, sans les mettre en commun avec les autres applications. En conséquence, ces dépendances sont installées isolément. Chaque application est autoportante.
+
+Si les dépendances sont mises en commun (avec les ` workspaces`), elles ne seront téléchargées et installées qu'une seule fois. Cela peut réduire le temps d'installation et la taille prise sur le disque. Cette solution n'est valide qu'en local, pas en production où les applications sont déployées séparément.
```
```suggestion
Si les dépendances sont mises en commun (avec les `workspaces`), elles ne seront téléchargées et installées qu'une seule fois. Cela peut réduire le temps d'installation et la taille prise sur le disque. Cette solution n'est valide qu'en local, pas en production où les applications sont déployées séparément.
```

yannbertrand commented:
Top l'ADR 👍

HEYGUL commented:
PR fermée car l'inconvénient d'avoir une seule et même version d'une dépendance au sein de **tout** le monorepo est trop important.
Décision consignée dans #7067.
